src/function/fetch.rs

```diff
@@ -83,7 +83,8 @@ where
         if let Some(memo) = memo_guard {
             let database_key_index = self.database_key_index(id);
             if memo.value.is_some()
-                && self.validate_may_be_provisional(db, zalsa, database_key_index, memo)
+                && (self.validate_may_be_provisional(db, zalsa, database_key_index, memo)
+                    || self.validate_same_iteration(db, database_key_index, memo))
```

Comment:
I suspect that the performance regression comes from `fetch_hot` now requiring at least two function calls if the `database_key_index` isn't provisional. Can you try marking `self.validate_same_iteration` as `#[cold]` so as to signal the compiler that this branch is unlikely.